### ISSUE TITLE
New Task: ATN DAC QC protocols and modified Argo QARTOD code, trajectory and profile are documented and shared via GitHub

### ISSUE THREAD

**Issue by MathewBiddle** [ioos/marine_life_data_network#14]
### Who is requesting this?

@marine-life

### What is being requested?

ATN DAC QC protocols and modified Argo QARTOD code, trajectory and profile are documented and shared via GitHub.

If this is something that builds on ioos_qc, enhancements should be proposed to that package.

### What is the requested deadline and why?

2024-07-01

### What is the current status quo (i.e., what happens if this does not get done)?

The Marine Life team and animal telemetry community will not fully understand what the QC protocols are for ATN data. 

### Provide a description or any other important information.

What indicates this is done?

Multiple actions could indicate this activity is done. The options are listed below but depend on what path the effort takes. These can be "either/or" but also "and" statements.

- [ ] A pull request is submitted to `ioos_qc` with the enhancements for modified Argo QARTOD
- [ ] The https://github.com/ioos/ioos-atn-data/ repository is updated to include a markdown describing QA/QC and modified Argo QARTOD
- [ ] The source code for modifications are published on GitHub at https://github.com/ioos/ioos-atn-data/, or other appropriate repository.

**Comment by mmckinzie** [comment]
Consider updating task title to: Document and share ATN DAC QC protocols. 
What is being requested?
- (Axiom/ML) Documentation about ATN DAC QC protocols, similar to: [Argo_CTD&TRAJ_qcmanual_version3.3.docx (noaa.gov)](https://cdn.ioos.noaa.gov/media/2020/03/Argo-QC-for-CTD-and-Trajectory-Data.pdf)

**Comment by MathewBiddle** [comment]
@mmckinzie it looks like [ioos_qc](https://github.com/ioos/ioos_qc) already had an argo test. See: 
- https://ioos.github.io/ioos_qc/api/ioos_qc.html#module-ioos_qc.argo 
- https://github.com/ioos/ioos_qc/blob/main/ioos_qc/argo.py

Please coordinate the ATN team with the ioos_qc package and submit changes upstream to ioos_qc. If you need help, please reach out.

**Comment by mmckinzie** [comment]
@MathewBiddle Yes, the Argo tests/code are well documented but modifications that were made to this base code and implemented by Axiom for the ATN DAC to make it applicable for animal borne platforms are not clearly documented anywhere. i.e., list of specific tests run for trajectory vs profile data, applied thresholds (different from those used for Argo floats), and modifications to base tests, e.g., date test includes not only impossible date_test but also whether or not location is within PI reported deployment window.

**Comment by mmckinzie** [comment]
Ian Jonsen is working on developing/updating IMOS near r/t QC protocols for animal borne profile data and corresponding trajectory files. This work is part of a GOOS Bio-Eco panel pilot project with AniBOS to serve behavioral data to OBIS/GBIF. Jon Pye will also be helping with this effort. Work to begin week of July 15th, 2024. This work will likely utilize elements of the [remora](https://github.com/ocean-tracking-network/remora/) package and OTN DwC templates/OBIS thematic node. This effort will be discussed at our upcoming AniBOS data committee meeting later this month.

I'll be curious to see how these IMOS/OTN protocols and QC procedures differ from what the ATN DAC is presently doing using QARTOD.

These new r/t QC tests will be distinct and differ from existing [MEOP](https://www.[meop.net/database/%5D(https://www.meop.net/database/)) protocols used to develop delay-mode data products, available here: https://osmc.noaa.gov/erddap/tabledap/MEOP_profiles.html

**Comment by MathewBiddle** [comment]
Thanks. What I read from that is there are three tasks for this:
1. Document the ATN modified Argo tests publicly
2. Compare the ATN modified Argo tests with the IMOS r/t QC protocols.
3. Do something with that comparison? Either adjust ATN's implementation or document the differences, I assume.

**Comment by mmckinzie** [comment]
Yes, I think that covers it.

**Comment by mmckinzie** [comment]
IMOS best practices manual for SMRU CTD data including instrument calibration, near real-time and delayed mode data QA/QC is now available: https://drive.google.com/file/d/1cwPNvqum4yXZUu1zfaY5TkSt_kHmPlhP/view?usp=sharing

**Comment by MathewBiddle** [comment]
Wooohoo! 

Do you know if it will be published in Ocean Best Practices? I would prefer to link to a DOI instead of a Google drive file.

**Comment by mmckinzie** [comment]
Yes, looks like it is already published on Ocean Best Practices -[ here](https://repository.oceanbestpractices.org/handle/11329/2571?show=full). DOI: https:doi.org/10.26198/ev75-0j83

**Comment by MathewBiddle** [comment]
How is what ATN is doing different (or the same) as https://github.com/ianjonsen/ArgosQC ?

**Comment by mmckinzie** [comment]
we only run the diag and ctd files through QARTOD

**Comment by MathewBiddle** [comment]
Has anyone started a repo for this effort? Should we be contributing back to the ArgosQC package?

**Comment by conniekot** [comment]
@iwensu0313 Please update this issue with progress, if it has been incorporated into https://github.com/ioos/ioos_qc, and any links to the code - thanks!

**Comment by iwensu0313** [comment]
Regarding **Documenting ATN DAC QC Protocols**, here are the QC configurations (I will want to double confirm with the team - will edit this as needed - but this is a start to kick off any discussions; let me know if you have Qs):

Default Thresholds for Platforms

| Test |  Suspect  | Fail | 
|:-----|:--------:|------:|
| ioos_qc.argo.speed_test   | 8.0 m/s | 10.0 m/s |
| ioos_qc.qartod.gross_range_test (temperature)   |  n/a  |   if outside of [-2.5, 40] |
| ioos_qc.qartod.gross_range_test (salinity)  | n/a |  if outside of  [0.0, 41.0] |
| ioos_qc.qartod.location_test | n/a | if outside of [-180, -90, 180, 90] |

Specific to SMRU Profiles
| Test |  Suspect  | Fail |
|:-----|:--------:|------:|
| ioos_qc.qartod.gross_range_test (temperature)   |  n/a  |  if outside of [-6, 57] |
| ioos_qc.qartod.gross_range_test (salinity)  | n/a |  if outside of  [2, 42.0] |
| Axiom system (sea_water_electrical_conductivity S/m)  | n/a |  if outside of  [0,200] |
| Axiom system (z)  | n/a |  if outside of   [-10, 10994] |
| ioos_qc.axds.valid_range_test (start/end time) | n/a | if outside time bounds from ADR |

Specific to SMRU Trajectory
| Test |  Suspect  | Fail |
|:-----|:--------:|------:|
| Axiom system (z)  | n/a |  if outside of  [0,0] |
| ioos_qc.axds.valid_range_test (start/end time) | n/a | if outside time bounds from ADR |

Specific to WC Trajectory
| Test |  Suspect  | Fail |
|:-----|:--------:|------:|
| ioos_qc.axds.valid_range_test (start/end time) | n/a | if outside time bounds from WC XML `deployment/start/date`, `deployment/end/date` |

**Comment by MathewBiddle** [comment]
@iwensu0313 why did this get moved from In Progress to Blocked?

Thank you for sharing those configurations. These are the types of information we'd like shared on https://ioos.github.io/ioos-atn-data/. Probably a sub-page on that site for QA/QC or something. 

We put together https://github.com/ioos/ioos-atn-data/blob/gh-pages/CONTRIBUTING.md#website which should be enough to get folks up to speed on adding content to the website. 

Were there any adjustments that were needed for the base `ioos_qc` package? Was the base `ioos_qc` package sufficient with these configurations?

**Comment by iwensu0313** [comment]
Moved to Blocked, unless we decide it is a priority for the upcoming 3 months (Dec 2024 - Feb 2025). Though a status like Paused would be more reflective.

> Thank you for sharing those configurations. These are the types of information we'd like shared on https://ioos.github.io/ioos-atn-data/. Probably a sub-page on that site for QA/QC or something.

Great, that's helpful confirmation! I do still want to double check with the team that these are right and up to date, while I'm pretty confident in my code navigating abilities :)


> Were there any adjustments that were needed for the base ioos_qc package? Was the base ioos_qc package sufficient with these configurations?

No adjustments to the `ioos_qc` package code base itself have been made from what I can tell. "Modifications" for ATN data are more on how we implement and the thresholds set; which I wouldn't really refer to as 'modifications'. For the most part, I think recent changes have been related to how we extracted the start/end date bounds to input into `ioos_qc.axds.valid_range_test`. These data points we grab from different places (ADR or the WC XML files). But maybe @mmckinzie has more to add here?

**Comment by MathewBiddle** [comment]
Awesome! Thank you for the additional details. 

Fantastic that the base `ioos_qc` package was sufficient for this purpose. 

Really, the goal of this activity was to fully describe what the QC protocols are for ATN and what packages are being used. So any non-technical person can understand what is being done.

So, in March we can add this documentation to https://ioos.github.io/ioos-atn-data/ for folks to have for reference. This needs to be in plain language.

**Comment by iwensu0313** [comment]
Moved from Blocked to In Progress to pick back up where I left off. Discussed at Dec 2025 Marine Life kickoff that documentation is an immediate priority.

I'll start looking at posting a PR with the information requested in https://ioos.github.io/ioos-atn-data/

**Comment by iwensu0313** [comment]
> A pull request is submitted to [ioos_qc](https://github.com/ioos/ioos_qc) with the enhancements for modified Argo QARTOD

The only "enhancement" I see is the following:

```
# Rename some columns until ioos_qc supports the stream_label attribute
        # in the configs
        renames = {
            'lon_qartod_location_test': 'qartod_location_test',
            'lon_argo_speed_test': 'argo_speed_test',
        }
        full_df = full_df.rename(columns=renames)
```

Other supplementary workflows primarily involve grabbing the metadata from different systems (e.g. grabbing start/end dates from the ADR instead of the data or threshold overrides in our internal platforms system that get put into the config passed into the ioos_qc functions). If there is a specific enhancement we are expecting I can try to look for it but I didn't see any.

Can ask the team to look into the stream_label attribute. Though doesn't seem high priority

**Comment by iwensu0313** [comment]
In progress in a [fork](https://github.com/iwensu0313/ioos-atn-data/commits/gh-pages/). For now I am adding a QC section to the end of the "ATN Satellite Telemetry Specification Version 1.0" page. Though it's quite long, it seemed to belong there. Can move to another section if we decide it needs to standalone.

**Comment by MathewBiddle** [comment]
@iwensu0313 This is great! Do you mind starting a draft Pull Request, so we can see what it is you added? LMK if you need help with that. Thanks!

**Comment by iwensu0313** [comment]
Please see https://github.com/ioos/ioos-atn-data/pull/108

We're also getting started on an example notebook. Where should that go? `_docs/notebooks`?

**Comment by MathewBiddle** [comment]
Yes, please!

**Comment by MathewBiddle** [comment]
@mmckinzie From the conversation the other week, my understanding is ATN is not QC'ing the animal tracks, instead QC'ing the environmental data as documented by @iwensu0313. If that's true (or whatever is the truth) can you add a sentence or two describing that process, in plain language, to: https://ioos.github.io/ioos-atn-data/atn-sat-telem-specification-v1-0.html#quality-control-protocols

The section to edit is at https://github.com/ioos/ioos-atn-data/blob/69b2fed361932ffa4a344073a47c6e6aaf8e4ff8/_docs/atn-sat-telem-specification-v1-0.md?plain=1#L1811

This will help us understand how to communicate the functions of the DAC more appropriately. 

Let me know if you have trouble editing the doc.

**Comment by mmckinzie** [comment]
@MathewBiddle We are QCing all auto-ingested data, both trajectory (tracks) and profile (sensor) data. However, a different set of tests is conducted depending on whether the trajectory or profile data are being QC'd. The QARTOD flags are embedded in the data files and are included in the standardized trajectory data we provide to NCEI, and they will also be included in future standardized profile data files.

Trajectory data = Impossible date test, impossible location test, and impossible speed test
Profile data = Impossible date test, impossible locations test, positions on land, impossible speed test, global range test, pressure increasing test, spike test

These QARTOD QC flags are not currently translated into the appropriate GTSPP quality flags (see below) in the BUFR messages; they are all hard-coded to 0 = unqualified. A[ solution](https://github.com/axiom-data-science/bufrtools/pull/12/changes/24b554802460c5615a17f19729f1757422a1542d#top) was proposed but never implemented.     

0 33 050 Global GTSPP quality flag Code table: 0 = Unqualified, 1 = Correct value (all checks passed), 2 = Probably good but value inconsistent with statistics (differ from climatology), 3 = Probably bad (spike, gradient, etc., if other tests passed), 4 = Bad value, impossible value (out of scale, vertical instability, constant profile), 5 = Value modified during quality control, 6–7 Reserved, 8 = Interpolated value, 9 = Good for operational use; caution; check literature for other uses, 10–14 Reserved, 15 = Missing value

Yes, I'll add it to the to-do list to include a few sentences in plain language in the protocols linked above.

**Comment by iwensu0313** [comment]
Thanks for the additional context @mmckinzie! I'll double check w/ the team that the current implementation in the system lines up w/ what you've described above.

**Comment by iwensu0313** [comment]
Hi @mmckinzie @MathewBiddle - an update on our review. 

The impossible date test, impossible location test, and impossible speed test are implemented for trajectory and profile data. An additional test - the global/gross range test - is applied for profile data if temp and salinity data are present. This aligns w/ what I documented in the [IOOS ATN QC protocols section](https://github.com/ioos/ioos-atn-data/pull/108/changes)

Confirming that we are not yet implementing the following:
- positions on land; test does not exist in ioos_qc, but we can develo
- pressure increasing test; test exists, can implement
- spike test; test exists, can implement (challenge will be if timeseries is irregularly spaced)

**Comment by MathewBiddle** [comment]
Thanks, @iwensu0313. Sounds like this issue can probably be closed once @mmckinzie updates the docs with that additional context.

Thanks for the heads up on the GTSPP quality flags. Is that something we need to be doing for the BUFR conversion and getting data to the GTS? Or, is it just additional context?

**Comment by iwensu0313** [comment]
A couple additional notes and updates
- We used to have an Argos pipeline which didn't have QC implemented at the time. I don't believe we ingest directly from Argos anymore but rather directly from WC/SMRU which do have QC. If this ever changes, we should revisit.
- Need to add table of contents section for QC in the left side bar (see screenshot)
- We mostly finished the trajectory section of the QC notebook example, but are finalizing the profile part (draft still in my forked branch not yet ready for review)

<img width="1343" height="565" alt="Image" src="https://github.com/user-attachments/assets/625c0df6-b329-421c-9aad-2e1c0e6698fc" />

**Comment by MathewBiddle** [comment]
Thanks @iwensu0313. Yeah, I was wondering if we should create a new page for the QA/QC documentation. Maybe list it under ATN Data Flows dropdown?

**Comment by iwensu0313** [comment]
I can move it to ATN Data Flow. working on finalizing the notebook example and pushing as PR for review soon

**Comment by iwensu0313** [comment]
Please see PR: https://github.com/ioos/ioos-atn-data/pull/110


We drafted a profile QC notebook example as well not yet pushed anywhere in github, but there are currently some slight road blocks, so I was thinking of revisiting the notebook example after some things have been more finmalized. In the trajectory notebook example, it was nice to be able to pull directly from a publicly available source for comparison (e.g. ATN Portal, NCEl). But we currently don't really surface _processed_ profile data anywhere yet even though it is technically run. As a reminder, which I will add to the DAC Overview documentation, it will be surfaced in two places once we finalize the following:
- We have this old PR that will help map the `ioos_qc` flags to the BUFR required format https://github.com/axiom-data-science/bufrtools/pull/12/changes, I believe, so that we can insert quality flags into the BUFR messages packaged and submitted to NDBC. 
- QCed Profile to NCEI

**Comment by MathewBiddle** [comment]
> send announcement sent through IOOS Eyes on the Ocean.

I'll add this to our biweekly. @laurabrenskelle FYI

**Comment by MathewBiddle** [comment]
Added to biweekly

**Comment by mmckinzie** [comment]
@iwensu0313 Thanks for all of your work on getting the QC procedures documented on the ATN specification website page under data flows. I know there is a way to do it, but how do you update content on those pages? I'd like to update the language about the valid range test under both the Quality Control Procedures and Deployment Duration sections; as saying that the valid range test is the only test that truncates the raw data based on provided deployment dates makes it sound like those positions are removed from all subsequent versions of the data. Which of course is not the case, positions are flagged as bad (4) and retained in most the processed data files (e.g., deployment.csv, NCEI .nc files, OBIS and ERDDAP files), while removed from aniMotum input and output files as well as used to truncate portal visualizations. It also not strictly positions on land that may fall outside of the reported deployment window, while this is often the case, it is not always the case.   

I also think it would be a good idea to include a simple table or text about the QARTOD flags themselves, so anyone viewing data files will know that 1 = good, 2 = not evaluated, 3 = suspect, 4 = bad/fail and 9 = missing data.   

The ATN DAC Flow Overview link is also giving a 404 error

**Comment by MathewBiddle** [comment]
This page? https://ioos.github.io/ioos-atn-data/overview.html

**Comment by MathewBiddle** [comment]
@mmckinzie If there are additional changes/updates you'd like to make to the docs, let's start a new ticket. The goal of this issue was to publish something on GitHub. Let's declare success and we can iterate on changes in another activity.

**Comment by mmckinzie** [comment]
Yes, the data overview page. The 404 error on the QC page is likely because the [data overview](https://ioos.github.io/ioos-atn-data/overview.html) page is still under development. 

Sounds good to start a new ticket for additional changes/updates to the [quality control protocols](https://ioos.github.io/ioos-atn-data/quality-control.html) page. I'll draft some updated language and share it with Iwen during our next coordinator call.

**Comment by MathewBiddle** [comment]
Ahh, I see now. The url was wrong on the QC page. I've updated it. Should be good now.

https://github.com/ioos/ioos-atn-data/commit/19aa515a1c722e5346ec42a50f2a14fb88253407
